### PR TITLE
48: Added 2 'For N in every Y items' reward expressions

### PR DESCRIPTION
**Pull request by yecli**
The new 'For N in every Y items' expressions to promotion expression tree #48:
Added required functionality

### CODE REVIEW

**Comment by yecli** [comment]
Screenshots:
Combo
rewardcombo
1 product reward
forevery1product
2 products reward
forevery2products

**Comment by yecli** [comment]
Reduced template strings:
image

File: VirtoCommerce.DynamicExpressionsModule.Data/Promotion/Rewards/RewardItemForEveryNumOtherItemInGetOfRel.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+namespace VirtoCommerce.DynamicExpressionsModule.Data.Promotion
+{
+    //For [] items of entry [] in every [] items of entry [] get [] % off no more than [] not to exceed $ []
+    public class RewardItemForEveryNumOtherItemInGetOfRel : RewardItemForEveryNumInGetOfRel
+    {
+        public ProductContainer ConditionalProduct { get; set; } = new ProductContainer();
+
+        #region IRewardExpression Members
+
+        public new PromotionReward[] GetRewards()
```
Comment (tatarincev) [inline VirtoCommerce.DynamicExpressionsModule.Data/Promotion/Rewards/RewardItemForEveryNumOtherItemInGetOfRel.cs]: Could you explain why do you use the `new` modificator here? It is absolutely  doesn't clear for me!

Comment (yecli) [reply VirtoCommerce.DynamicExpressionsModule.Data/Promotion/Rewards/RewardItemForEveryNumOtherItemInGetOfRel.cs]: It is a mistake in fact. Should be override instead. Will change

File: VirtoCommerce.DynamicExpressionsModule.Data/Promotion/Rewards/RewardItemForEveryNumOtherItemInGetOfRel.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    //For [] items of entry [] in every [] items of entry [] get [] % off no more than [] not to exceed $ []
+    public class RewardItemForEveryNumOtherItemInGetOfRel : RewardItemForEveryNumInGetOfRel
+    {
+        public ProductContainer ConditionalProduct { get; set; } = new ProductContainer();
+
+        #region IRewardExpression Members
+
+        public override PromotionReward[] GetRewards()
+        {
+            return new PromotionReward[] {
```
Comment (tatarincev) [inline VirtoCommerce.DynamicExpressionsModule.Data/Promotion/Rewards/RewardItemForEveryNumOtherItemInGetOfRel.cs]: Can we reuse reward object already created in base class  and just set ConditionalProduct to it?

Comment (yecli) [reply VirtoCommerce.DynamicExpressionsModule.Data/Promotion/Rewards/RewardItemForEveryNumOtherItemInGetOfRel.cs]: Yes, we can. Did not make it that way initially as in that case we would have a tight dependency on base class implementation. Thought about this again - probably we want the same values to be filled with the same fields in the same order in our case, so yes, better to use base method.

File: VirtoCommerce.DynamicExpressionsModule.Data/Promotion/Rewards/RewardItemForEveryNumInGetOfRel.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        public int ForNthQuantity { get; set; }
+        public int InEveryNthQuantity { get; set; }
+        public int ItemLimit { get; set; }
+        public decimal MaxLimit { get; set; }
+
+        #region IRewardExpression Members
+
+        public PromotionReward[] GetRewards()
+        {
+            var reward = new CatalogItemAmountReward();
```
Comment (tatarincev) [inline VirtoCommerce.DynamicExpressionsModule.Data/Promotion/Rewards/RewardItemForEveryNumInGetOfRel.cs]: I suggest to do not increase complicity by adding the new method
just write this code
```
var result = base.GetRewards().FirstOrDefault();
if(result != null)
{
result.ConditionalProductId = ConditionalProduct?.ProductId;

}
return result;
```

Comment (yecli) [reply VirtoCommerce.DynamicExpressionsModule.Data/Promotion/Rewards/RewardItemForEveryNumInGetOfRel.cs]: We have method GetRewards() returning PromotionReward[] array, so we would need a cast in child object. In that case we would have non testable in compile time agreement for base class return type.
I think strictly typed method is better, so the complexity is reasonable here. 
What do yo think?

Comment (tatarincev) [reply VirtoCommerce.DynamicExpressionsModule.Data/Promotion/Rewards/RewardItemForEveryNumInGetOfRel.cs]: But you directly inherit from the class where you created  this return object!

Comment (yecli) [reply VirtoCommerce.DynamicExpressionsModule.Data/Promotion/Rewards/RewardItemForEveryNumInGetOfRel.cs]: Yes, but if you change the type of created reward in base class, you probably will have an error (or undesired behavior if using "as") for the child in runtime. Here i could see implicit dependency, which could be eliminated using this additional method. I think it is safer.
But OK, i could change it.

**Pull request merged**